src/main/java/com/squareup/phrase/ListPhrase.java

```diff
@@ -0,0 +1,200 @@
+    return join(asList(first, second, rest));
+  }
+
+  public <T> CharSequence join(Iterable<T> items) {
+    checkNotNull(items);
+    return join(items, null);
+  }
+
+  public <T> CharSequence join(Iterable<T> items, Formatter<T> formatter) {
+    checkNotNull(items);
+    return joinIterableWithSize(items, getSize(items), formatter);
+  }
+
+  private <T> CharSequence joinIterableWithSize(Iterable<T> items, int size, Formatter<T> formatter) {
+    switch (size) {
+      case 0:
+        return "";
+      case 1:
+        return formatIfNotNull(items.iterator().next(), formatter);
+      case 2:
+        return joinTwoElements(items, formatter);
+      default:
+        return joinMoreThanTwoElements(items, size, formatter);
+    }
+  }
+
+  private <T> CharSequence joinTwoElements(Iterable<T> items, Formatter<T> formatter) {
+    StringBuilder builder = new StringBuilder();
+    Iterator<T> iterator = items.iterator();
+
+    // Don't need to check hasNext since we know the size.
+    builder.append(formatIfNotNull(iterator.next(), formatter));
+    builder.append(twoElementSeparator);
+    builder.append(formatIfNotNull(iterator.next(), formatter));
+    return builder.toString();
+  }
+
+  private <T> CharSequence joinMoreThanTwoElements(Iterable<T> items, int size, Formatter<T> formatter) {
+    StringBuilder builder = new StringBuilder();
+    int secondLastIndex = size - 2;
+    Iterator<T> iterator = items.iterator();
+
+    for (int i = 0; i < size; i++) {
+      // Don't need to check hasNext since we know the size.
+      builder.append(formatIfNotNull(iterator.next(), formatter));
+
+      if (i < secondLastIndex) {
+        builder.append(nonFinalElementSeparator);
+      } else if (i == secondLastIndex) {
+        builder.append(finalElementSeparator);
+      }
+    }
+
+    return builder.toString();
+  }
+
+  private static int getSize(Iterable<?> iterable) {
+    if (iterable instanceof Collection) {
+      return ((Collection) iterable).size();
+    }
+
+    int size = 0;
+    Iterator<?> it = iterable.iterator();
+    while (it.hasNext()) {
+      size++;
+      it.next();
+    }
+    return size;
+  }
+
+  private static <T> List<T> asList(final T first, final T second, final T[] rest) {
+    return new AbstractList<T>() {
+      @Override public T get(int index) {
+        switch (index) {
+          case 0:
+            return first;
+          case 1:
+            return second;
+          default:
+            return rest[index-2];
+        }
+      }
+
+      @Override public int size() {
+        return rest.length + 2;
+      }
+    };
+  }
+
+
+  private static <T> CharSequence formatIfNotNull(T item, Formatter<T> formatter) {
+    if (formatter == null) {
+      return item == null ? "" : item.toString();
+    }
+
+    return formatter.format(item);
+  }
+
+  private static void checkNotNull(Object obj) {
+    if (obj == null) {
+      throw new IllegalArgumentException("list cannot be null");
```

Comment:
Kinda weird to take an Object and then complain about a list.  Our usual `checkNotNull` takes a message parameter for this reason.